cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -357,9 +357,8 @@ protected void shutDown() throws Exception {
       try {
         if (mapReduce instanceof ProgramLifecycle) {
           destroy(success, failureInfo);
-        } else {
-          onFinish(success);
         }
+        onFinish(success);
```

Comment:
@chtyim Per discussion, made the following changes:
1. Made `initialize(context)` method as final.
2. Added another version of the initialize method which takes no argument but can be extended by User program.
3. MapReduceRuntimeService calls the `initialize(context)` and no arg `initialize` methods if its instance of ProgramLifecycle else it calls `beforeSubmit`